x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java

```diff
@@ -0,0 +1,276 @@
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xpack.core.security.authc.support.mapper.ExpressionRoleMapping;
+import org.elasticsearch.xpack.core.security.authc.support.mapper.expressiondsl.FieldExpression;
+import org.elasticsearch.xpack.security.SecurityOnTrialLicenseRestTestCase;
+import org.junit.ClassRule;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+
+public class RoleMappingRestIT extends ESRestTestCase {
+    private static final String settingsJson = """
+        {
+             "metadata": {
+                 "version": "1",
+                 "compatibility": "8.4.0"
+             },
+             "state": {
+                 "role_mappings": {
+                       "role-mapping-1": {
+                          "enabled": true,
+                          "roles": [ "role_1" ],
+                          "rules": { "field": { "username": "no_user" } },
+                          "metadata": {
+                             "uuid" : "b9a59ba9-6b92-4be2-bb8d-02bb270cb3a7",
+                             "_foo": "something",
+                             "_es_reserved_role_mapping_name": "ignored"
+                          }
+                       },
+                       "role-mapping-2": {
+                          "enabled": true,
+                          "roles": [ "role_2" ],
+                          "rules": { "field": { "username": "no_user" } }
+                       },
+                       "role-mapping-3": {
+                          "enabled": true,
+                          "roles": [ "role_3" ],
+                          "rules": { "field": { "username": "no_user" } },
+                          "metadata": {
+                             "_read_only" : { "field": 1 },
+                             "_es_reserved_role_mapping_name": { "still_ignored": true }
+                          }
+                       }
+                 }
+             }
+        }""";
+    private static final ExpressionRoleMapping clusterStateMapping1 = new ExpressionRoleMapping(
+        "role-mapping-1-read-only-operator-mapping",
+        new FieldExpression("username", List.of(new FieldExpression.FieldValue("no_user"))),
+        List.of("role_1"),
+        null,
+        Map.of("uuid", "b9a59ba9-6b92-4be2-bb8d-02bb270cb3a7", "_foo", "something", "_read_only", true),
+        true
+    );
+    private static final ExpressionRoleMapping clusterStateMapping2 = new ExpressionRoleMapping(
+        "role-mapping-2-read-only-operator-mapping",
```

Comment:
I'm going with the `-read-only-operator-mapping` because it got too depressing writing the REST tests with the alternative: they don't work with spaces unless you URL-encode the path. 

If we want to make for nicer display items in Kibana, we could always render the role mapping names different there whereas we can't do anything about spaces in the ES API.

Reply:
Thanks for raising this!

Fully agreed that it looks better in the Kibana role mapping page. Dev Tools however becomes very difficult to use because the path is also separated by commas. Furthermore, switching to ` (read-only)` still results in fairly confusing test failures even with URL encoding. For instance the equality checks for the PUT API don't work anymore. Handling URL encoding feels error-prone and not worth the effort.

> If you decide to not do that, I think including the original name in the metadata returned would be nice so Kibana doesn't have to do some string magic to strip the suffix.

I think the right answer might be to add a new field called `display_name` to the top-level expression mapping XContent class. I'm really hesitant to use a metadata field beyond the backend hack (which is still something we can choose to clean up) for something that really feels like a missing top-level field. Or better yet: at the XContent layer, we could add an `id` field, and let `name` become "display name". I don't think any of this (not even surfacing the metadata field in Kibana) is viable for 8.16.0 though so I propose to go with the current implementation, and then follow up, having coordinated with Kibana.

Does this sound like an okay path forward?

Reply:
Yes, that sounds good!  Thanks for exploring this path. I don't think it's worth spending too much time getting this to work. 

I like the idea of a display name eventually, but like you said, not something we should add until there is a need for it. Adding it to the metadata might force us to have it there forever, so I agree that it makes sense to not do that at this point.